stack/create.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stack
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/mineiros-io/terramate/errors"
+	"github.com/mineiros-io/terramate/hcl"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	// ErrInvalidStackDir indicates that the given stack dir is invalid.
+	ErrInvalidStackDir errors.Kind = "invalid stack directory"
+
+	// ErrInvalidStackID indicates that the given stack ID is invalid.
+	ErrInvalidStackID errors.Kind = "invalid stack ID"
+
+	// ErrStackAlreadyExists indicates that the stack already exists and cant be created.
+	ErrStackAlreadyExists errors.Kind = "stack already exists"
+)
+
+// CreateCfg represents stack creation configuration.
+type CreateCfg struct {
+	// Dir is the absolute path of the directory, inside the project root,
+	// where the stack will be created. It must be non-empty.
+	Dir string
+
+	// ID of the stack, defaults to an UUID V4 if empty.
+	ID string
+
+	// Name of the stack, defaults to Dir basename if empty.
+	Name string
+
+	// Description of the stack, defaults to Name if empty.
+	Description string
+
+	// Imports represents a set of import paths.
+	Imports []string
+}
+
+// Create creates a stack according to the given configuration.
+// Any dirs on the path provided on the configuration that doesn't exist
+// will be created.
+//
+// If the stack already exists it will return an error and no changes will be
+// made to the stack.
+func Create(rootdir string, cfg CreateCfg) error {
+	const stackFilename = "stack.tm.hcl"
+
+	logger := log.With().
+		Str("action", "stack.Create()").
+		Stringer("cfg", cfg).
+		Logger()
+
+	if !filepath.IsAbs(cfg.Dir) {
+		return errors.E(ErrInvalidStackDir, "relative paths not allowed")
+	}
+
+	if !strings.HasPrefix(cfg.Dir, rootdir) {
+		return errors.E(ErrInvalidStackDir, "stack %q must be inside project root %q", cfg.Dir, rootdir)
+	}
+
+	if strings.HasPrefix(filepath.Base(cfg.Dir), ".") {
+		return errors.E(ErrInvalidStackDir, "dot directories not allowed")
+	}
+
+	logger.Trace().Msg("creating stack dir if absent")
+
+	if err := os.MkdirAll(cfg.Dir, 0755); err != nil {
+		return errors.E(err, "failed to create new stack directories")
+	}
+
+	logger.Trace().Msg("validating create configuration")
+
+	_, err := os.Stat(filepath.Join(cfg.Dir, stackFilename))
+	if err == nil {
+		// Even if there is no stack inside the file, we can't overwrite
+		// the user file anyway.
+		return errors.E(ErrStackAlreadyExists, "%q already exists", stackFilename)
+	}
+
+	// We could have a stack definition somewhere else.
+	parsedCfg, err := hcl.ParseDir(rootdir, cfg.Dir)
+	if err != nil {
+		return errors.E(err, "invalid config creating stack dir %s", cfg.Dir)
+	}
+
+	if parsedCfg.Stack != nil {
+		return errors.E(ErrStackAlreadyExists, "name %q", parsedCfg.Stack.Name)
+	}
+
+	hclCfg, err := hcl.NewConfig(cfg.Dir)
```

Comment:
`hclCfg` is the **terramate configuration** (everything) and `cfg` is the **stack creation configuration**. The naming is a bit confusing because in the end both are hcl related configs. What about:

- tmcfg (terramate config)
- cfg (stack config - the context fills the rest.. package is stack, etc)

Reply:
True true, I named initially after its origin from the hcl package :laughing: